File: _product_of_array_except_self.ts
function productExceptSelf(nums: number[]): number[] {
    let res = []
    let preArr = []
    let postArr = []
    let pre = 1;
    let post = 1;

    for(let i =0; i<nums.length; i++){
        let temp = pre * nums[i]
        preArr.push(temp)
        pre=temp
    }

    for(let i=nums.length-1; i>=0; i--){
        let temp = post*nums[i]
        postArr.unshift(temp)
        post=temp
    }

    res.push(postArr[1])

    for(let i =1; i<nums.length-1;i++){
        let temp = preArr[i-1] * postArr[i+1]
        res.push(temp)
    }

    res.push(preArr[nums.length-2])

    return res;

};

File: Python/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        left = []
        
        pre =1
        post = 1
        
        n=len(nums)
        right = [0]*(n)

        res = [0] *n
        

        for i in range(0,n):
            val = pre*nums[i]
            pre=val
            left.append(val)
        
        for i in range(n-1,-1,-1):
            val = post*nums[i]
            post = val
            right[i] = val
        
        # print('left',left)
        # print('right',right)

        res[0] = right[1]
        res[n-1] = left[n-2]

        for i in range(1, n-1):
            res[i] = right[i+1] * left[i-1]

        return res
